File: packages/scripts/download-packages.js
#!/usr/bin/env node

/**
 * Package Download Automation System
 * Handles downloading packages from remote registries with caching and dependency tracking
 */

const fs = require('fs').promises;
const path = require('path');
const https = require('https');
const crypto = require('crypto');

class PackageDownloader {
    constructor(configPath = '../package-registry.json') {
        this.configPath = configPath;
        this.config = null;
        this.cacheDir = path.join(__dirname, '../cache');
        this.downloadQueue = [];
    }

    async initialize() {
        try {
            const configData = await fs.readFile(path.join(__dirname, this.configPath), 'utf8');
            this.config = JSON.parse(configData);
            
            // Ensure cache directory exists
            await fs.mkdir(this.cacheDir, { recursive: true });
            
            console.log('Package downloader initialized successfully');
        } catch (error) {
            throw new Error(`Failed to initialize package downloader: ${error.message}`);
        }
    }

    /**
     * Download package from remote registry
     */
    async downloadPackage(packageName, version = 'latest', registry = 'npm') {
        try {
            const registryConfig = this.config.registries[registry];
            if (!registryConfig || !registryConfig.enabled) {
                throw new Error(`Registry ${registry} is not available or disabled`);
            }

            console.log(`Downloading ${packageName}@${version} from ${registry}...`);

            const packageInfo = await this.getPackageInfo(packageName, version, registryConfig);
            const downloadUrl = this.buildDownloadUrl(packageInfo, registryConfig);
            
            // Check cache first
            const cacheKey = this.generateCacheKey(packageName, version, registry);
            const cachedPath = await this.checkCache(cacheKey, registryConfig.cache_ttl);
            
            if (cachedPath) {
                console.log(`Using cached version of ${packageName}@${version}`);
                return cachedPath;
            }

            // Download package
            const downloadPath = await this.performDownload(downloadUrl, packageName, version);
            
            // Cache the downloaded package
            if (registryConfig.cache) {
                await this.cachePackage(cacheKey, downloadPath);
            }

            // Track dependency
            await this.trackDependency(packageName, version, registry, downloadPath);

            console.log(`Successfully downloaded ${packageName}@${version}`);
            return downloadPath;

        } catch (error) {
            console.error(`Failed to download ${packageName}@${version}: ${error.message}`);
            throw error;
        }
    }

    /**
     * Get package information from registry
     */
    async getPackageInfo(packageName, version, registryConfig) {
        // This would implement registry-specific API calls
        // For now, return mock data
        return {
            name: packageName,
            version: version,
            downloadUrl: `${registryConfig.url}${packageName}/-/${packageName}-${version}.tgz`,
            dependencies: {},
            devDependencies: {}
        };
    }

    /**
     * Build download URL based on registry type
     */
    buildDownloadUrl(packageInfo, registryConfig) {
        if (registryConfig.type === 'npm') {
            return packageInfo.downloadUrl;
        } else if (registryConfig.type === 'pypi') {
            return `${registryConfig.url}${packageInfo.name}/`;
        }
        return packageInfo.downloadUrl;
    }

    /**
     * Generate cache key for package
     */
    generateCacheKey(packageName, version, registry) {
        const input = `${packageName}-${version}-${registry}`;
        return crypto.createHash('sha256').update(input).digest('hex');
    }

    /**
     * Check if package exists in cache and is not expired
     */
    async checkCache(cacheKey, ttl) {
        try {
            const cachePath = path.join(this.cacheDir, cacheKey);
            const stats = await fs.stat(cachePath);
            
            if (ttl && Date.now() - stats.mtime.getTime() > ttl * 1000) {
                return null; // Expired
            }
            
            return cachePath;
        } catch (error) {
            return null; // Not in cache
        }
    }

    /**
     * Perform actual download
     */
    async performDownload(url, packageName, version) {
        return new Promise((resolve, reject) => {
            const fileName = `${packageName}-${version}.tgz`;
            const downloadPath = path.join(__dirname, '../remote', fileName);
            
            // Ensure remote directory exists
            fs.mkdir(path.dirname(downloadPath), { recursive: true }).then(() => {
                const file = require('fs').createWriteStream(downloadPath);
                
                https.get(url, (response) => {
                    if (response.statusCode !== 200) {
                        reject(new Error(`HTTP ${response.statusCode}: ${response.statusMessage}`));
                        return;
                    }
                    
                    response.pipe(file);
                    
                    file.on('finish', () => {
                        file.close();
                        resolve(downloadPath);
                    });
                    
                    file.on('error', (error) => {
                        fs.unlink(downloadPath);
                        reject(error);
                    });
                }).on('error', reject);
            });
        });
    }

    /**
     * Cache downloaded package
     */
    async cachePackage(cacheKey, sourcePath) {
        const cachePath = path.join(this.cacheDir, cacheKey);
        await fs.copyFile(sourcePath, cachePath);
    }

    /**
     * Track package dependency
     */
    async trackDependency(packageName, version, registry, downloadPath) {
        if (!this.config.dependency_tracking.enabled) {
            return;
        }

        const trackingFile = path.join(__dirname, '../dependency-tracking.json');
        let tracking = {};

        try {
            const trackingData = await fs.readFile(trackingFile, 'utf8');
            tracking = JSON.parse(trackingData);
        } catch (error) {
            // File doesn't exist, start fresh
        }

        if (!tracking.dependencies) {
            tracking.dependencies = {};
        }

        tracking.dependencies[packageName] = {
            version,
            registry,
            downloadPath,
            downloadDate: new Date().toISOString(),
            lastUsed: new Date().toISOString()
        };

        await fs.writeFile(trackingFile, JSON.stringify(tracking, null, 2));
    }

    /**
     * Batch download packages
     */
    async downloadBatch(packages) {
        const results = [];
        
        for (const pkg of packages) {
            try {
                const result = await this.downloadPackage(
                    pkg.name, 
                    pkg.version || 'latest', 
                    pkg.registry || 'npm'
                );
                results.push({ ...pkg, success: true, path: result });
            } catch (error) {
                results.push({ ...pkg, success: false, error: error.message });
            }
        }
        
        return results;
    }

    /**
     * Clean expired cache entries
     */
    async cleanCache() {
        try {
            const files = await fs.readdir(this.cacheDir);
            let cleaned = 0;

            for (const file of files) {
                const filePath = path.join(this.cacheDir, file);
                const stats = await fs.stat(filePath);
                
                // Remove files older than 7 days
                if (Date.now() - stats.mtime.getTime() > 7 * 24 * 60 * 60 * 1000) {
                    await fs.unlink(filePath);
                    cleaned++;
                }
            }

            console.log(`Cleaned ${cleaned} expired cache entries`);
        } catch (error) {
            console.error(`Failed to clean cache: ${error.message}`);
        }
    }
}

// CLI interface
if (require.main === module) {
    const downloader = new PackageDownloader();
    
    async function main() {
        const args = process.argv.slice(2);
        
        if (args.length === 0) {
            console.log('Usage: node download-packages.js <command> [options]');
            console.log('Commands:');
            console.log('  download <package> [version] [registry] - Download a single package');
            console.log('  batch <file>                           - Download packages from JSON file');
            console.log('  clean                                  - Clean expired cache entries');
            return;
        }

        await downloader.initialize();

        const command = args[0];
        
        switch (command) {
            case 'download':
                if (args.length < 2) {
                    console.error('Package name required');
                    return;
                }
                await downloader.downloadPackage(args[1], args[2], args[3]);
                break;
                
            case 'batch':
                if (args.length < 2) {
                    console.error('Batch file required');
                    return;
                }
                const batchFile = await fs.readFile(args[1], 'utf8');
                const packages = JSON.parse(batchFile);
                const results = await downloader.downloadBatch(packages);
                console.log('Batch download results:', results);
                break;
                
            case 'clean':
                await downloader.cleanCache();
                break;
                
            default:
                console.error(`Unknown command: ${command}`);
        }
    }

    main().catch(console.error);
}

module.exports = PackageDownloader;

File: packages/scripts/download-packages.py
#!/usr/bin/env python3

"""
Python Package Download Automation System
Handles downloading Python packages from PyPI and other registries with caching and dependency tracking
"""

import os
import sys
import json
import hashlib
import shutil
import requests
import subprocess
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Optional, Tuple

class PythonPackageDownloader:
    def __init__(self, config_path: str = "../package-registry.json"):
        self.config_path = config_path
        self.config = None
        self.cache_dir = Path(__dirname__).parent / "cache"
        self.requirements_cache = {}
        
    async def initialize(self):
        """Initialize the package downloader"""
        try:
            config_file = Path(__file__).parent / self.config_path
            with open(config_file, 'r') as f:
                self.config = json.load(f)
                
            # Ensure cache directory exists
            self.cache_dir.mkdir(parents=True, exist_ok=True)
            
            print("Python package downloader initialized successfully")
        except Exception as e:
            raise Exception(f"Failed to initialize package downloader: {str(e)}")
    
    def download_package(self, package_name: str, version: str = "latest", 
                        registry: str = "pypi") -> str:
        """Download a Python package from registry"""
        try:
            registry_config = self.config["registries"][registry]
            if not registry_config or not registry_config["enabled"]:
                raise Exception(f"Registry {registry} is not available or disabled")
                
            print(f"Downloading {package_name}=={version} from {registry}...")
            
            # Check cache first
            cache_key = self._generate_cache_key(package_name, version, registry)
            cached_path = self._check_cache(cache_key, registry_config.get("cache_ttl"))
            
            if cached_path:
                print(f"Using cached version of {package_name}=={version}")
                return cached_path
                
            # Download using pip
            download_path = self._download_with_pip(package_name, version)
            
            # Cache the downloaded package
            if registry_config.get("cache"):
                self._cache_package(cache_key, download_path)
                
            # Track dependency
            self._track_dependency(package_name, version, registry, download_path)
            
            print(f"Successfully downloaded {package_name}=={version}")
            return download_path
            
        except Exception as e:
            print(f"Failed to download {package_name}=={version}: {str(e)}")
            raise e
    
    def _download_with_pip(self, package_name: str, version: str) -> str:
        """Download package using pip"""
        download_dir = Path(__file__).parent.parent / "remote" / "python"
        download_dir.mkdir(parents=True, exist_ok=True)
        
        if version == "latest":
            package_spec = package_name
        else:
            package_spec = f"{package_name}=={version}"
            
        # Use pip download to get the package
        cmd = [
            sys.executable, "-m", "pip", "download",
            "--dest", str(download_dir),
            "--no-deps",  # Don't download dependencies automatically
            package_spec
        ]
        
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f"pip download failed: {result.stderr}")
            
        # Find the downloaded file
        downloaded_files = list(download_dir.glob(f"{package_name.replace('-', '_')}*"))
        if not downloaded_files:
            raise Exception(f"Downloaded package not found in {download_dir}")
            
        return str(downloaded_files[0])
    
    def _generate_cache_key(self, package_name: str, version: str, registry: str) -> str:
        """Generate cache key for package"""
        input_str = f"{package_name}-{version}-{registry}"
        return hashlib.sha256(input_str.encode()).hexdigest()
    
    def _check_cache(self, cache_key: str, ttl: Optional[int]) -> Optional[str]:
        """Check if package exists in cache and is not expired"""
        try:
            cache_path = self.cache_dir / cache_key
            if not cache_path.exists():
                return None
                
            if ttl:
                mtime = datetime.fromtimestamp(cache_path.stat().st_mtime)
                if datetime.now() - mtime > timedelta(seconds=ttl):
                    return None  # Expired
                    
            return str(cache_path)
        except:
            return None
    
    def _cache_package(self, cache_key: str, source_path: str):
        """Cache downloaded package"""
        cache_path = self.cache_dir / cache_key
        shutil.copy2(source_path, cache_path)
    
    def _track_dependency(self, package_name: str, version: str, 
                         registry: str, download_path: str):
        """Track package dependency"""
        if not self.config["dependency_tracking"]["enabled"]:
            return
            
        tracking_file = Path(__file__).parent.parent / "dependency-tracking.json"
        tracking = {}
        
        try:
            if tracking_file.exists():
                with open(tracking_file, 'r') as f:
                    tracking = json.load(f)
        except:
            pass
            
        if "dependencies" not in tracking:
            tracking["dependencies"] = {}
            
        tracking["dependencies"][package_name] = {
            "version": version,
            "registry": registry,
            "downloadPath": download_path,
            "downloadDate": datetime.now().isoformat(),
            "lastUsed": datetime.now().isoformat(),
            "language": "python"
        }
        
        with open(tracking_file, 'w') as f:
            json.dump(tracking, f, indent=2)
    
    def download_batch(self, packages: List[Dict]) -> List[Dict]:
        """Download multiple packages"""
        results = []
        
        for pkg in packages:
            try:
                result = self.download_package(
                    pkg["name"],
                    pkg.get("version", "latest"),
                    pkg.get("registry", "pypi")
                )
                results.append({**pkg, "success": True, "path": result})
            except Exception as e:
                results.append({**pkg, "success": False, "error": str(e)})
                
        return results
    
    def install_from_requirements(self, requirements_file: str) -> Dict:
        """Install packages from requirements.txt file"""
        try:
            print(f"Installing packages from {requirements_file}")
            
            cmd = [
                sys.executable, "-m", "pip", "install",
                "-r", requirements_file,
                "--cache-dir", str(self.cache_dir)
            ]
            
            result = subprocess.run(cmd, capture_output=True, text=True)
            
            return {
                "success": result.returncode == 0,
                "stdout": result.stdout,
                "stderr": result.stderr
            }
            
        except Exception as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    def generate_requirements_lock(self, output_file: str = "requirements.lock"):
        """Generate locked requirements file with exact versions"""
        try:
            cmd = [sys.executable, "-m", "pip", "freeze"]
            result = subprocess.run(cmd, capture_output=True, text=True)
            
            if result.returncode == 0:
                output_path = Path(__file__).parent.parent / output_file
                with open(output_path, 'w') as f:
                    f.write(result.stdout)
                print(f"Generated locked requirements file: {output_path}")
                return str(output_path)
            else:
                raise Exception(f"pip freeze failed: {result.stderr}")
                
        except Exception as e:
            print(f"Failed to generate requirements lock: {str(e)}")
            raise e
    
    def clean_cache(self):
        """Clean expired cache entries"""
        try:
            cleaned = 0
            cutoff_time = datetime.now() - timedelta(days=7)
            
            for cache_file in self.cache_dir.iterdir():
                if cache_file.is_file():
                    mtime = datetime.fromtimestamp(cache_file.stat().st_mtime)
                    if mtime < cutoff_time:
                        cache_file.unlink()
                        cleaned += 1
                        
            print(f"Cleaned {cleaned} expired cache entries")
        except Exception as e:
            print(f"Failed to clean cache: {str(e)}")

def main():
    """CLI interface"""
    import asyncio
    
    if len(sys.argv) < 2:
        print("Usage: python download-packages.py <command> [options]")
        print("Commands:")
        print("  download <package> [version] [registry] - Download a single package")
        print("  batch <file>                           - Download packages from JSON file")
        print("  install <requirements_file>            - Install from requirements file")
        print("  freeze [output_file]                   - Generate locked requirements")
        print("  clean                                  - Clean expired cache entries")
        return
        
    downloader = PythonPackageDownloader()
    asyncio.run(downloader.initialize())
    
    command = sys.argv[1]
    
    if command == "download":
        if len(sys.argv) < 3:
            print("Package name required")
            return
        downloader.download_package(
            sys.argv[2],
            sys.argv[3] if len(sys.argv) > 3 else "latest",
            sys.argv[4] if len(sys.argv) > 4 else "pypi"
        )
        
    elif command == "batch":
        if len(sys.argv) < 3:
            print("Batch file required")
            return
        with open(sys.argv[2], 'r') as f:
            packages = json.load(f)
        results = downloader.download_batch(packages)
        print("Batch download results:", json.dumps(results, indent=2))
        
    elif command == "install":
        if len(sys.argv) < 3:
            print("Requirements file required")
            return
        result = downloader.install_from_requirements(sys.argv[2])
        if result["success"]:
            print("Installation completed successfully")
        else:
            print(f"Installation failed: {result.get('error', result.get('stderr'))}")
            
    elif command == "freeze":
        output_file = sys.argv[2] if len(sys.argv) > 2 else "requirements.lock"
        downloader.generate_requirements_lock(output_file)
        
    elif command == "clean":
        downloader.clean_cache()
        
    else:
        print(f"Unknown command: {command}")

if __name__ == "__main__":
    main()
